# CVE-2014-1912

## Experiment Environment

CentOS 6.5

## INSTALL & Configuration

```
wget https://github.com/mudongliang/source-packages/raw/master/CVE-2014-1912/Python-3.1.5.tgz
tar -xvf Python-3.1.5.tgz
cd Python-3.1.5
./configure
make
```

## Problems in Installation & Configuration


## How to trigger vulnerability

### Method 1

**Note:** it works both Python2 and Python3

```
./python poc1.py
```

**Note:** it works only in Python 2

### Metohd 2

```

./python poc2.py
cat egg | nc -l 8080 -vv
./python client.py
```

## PoCs

[buffer overflow in socket.recvfrom_into](https://bugs.python.org/issue20246)

[Python 'sock_recvfrom_into()' Function Buffer Overflow Vulnerability](https://www.securityfocus.com/bid/65379/exploit)

[Python - 'socket.recvfrom_into()' Remote Buffer Overflow](https://www.exploit-db.com/exploits/31875/)

## Vulnerability Details & Patch

### Root Cause

### Stack Trace

### Patch

```
diff -r e6358103fe4f Modules/socketmodule.c
--- a/Modules/socketmodule.c	Wed Jan 08 20:44:37 2014 -0800
+++ b/Modules/socketmodule.c	Sun Jan 12 13:21:19 2014 -0800
@@ -2877,6 +2877,14 @@
         recvlen = buflen;
     }
 
+    /* Check if the buffer is large enough */
+    if (buflen < recvlen) {
+        PyBuffer_Release(&pbuf);
+        PyErr_SetString(PyExc_ValueError,
+                        "buffer too small for requested bytes");
+        return NULL;
+    }
+
     readlen = sock_recvfrom_guts(s, buf, recvlen, flags, &addr);
     if (readlen < 0) {
         PyBuffer_Release(&pbuf);
```

## References
